File: lib/domain/field/dtos/field_create_input.rb
# frozen_string_literal: true

module Domain
  module Field
    module Dtos
      class FieldCreateInput
        attr_reader :name, :area, :daily_fixed_cost, :region, :farm_id

        def initialize(name:, farm_id:, area: nil, daily_fixed_cost: nil, region: nil)
          @name = name
          @farm_id = farm_id
          @area = area
          @daily_fixed_cost = daily_fixed_cost
          @region = region
        end

        def self.from_hash(hash, farm_id: nil)
          pp = hash[:field] || hash
          resolved_farm_id = farm_id || hash[:farm_id]
          new(
            name: pp[:name],
            farm_id: resolved_farm_id,
            area: pp[:area],
            daily_fixed_cost: pp[:daily_fixed_cost],
            region: pp[:region]
          )
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field/dtos/field_create_input.rs
/// Ruby: `Domain::Field::Dtos::FieldCreateInput`
#[derive(Debug, Clone, PartialEq)]
pub struct FieldCreateInput {
    pub name: String,
    pub farm_id: i64,
    pub area: Option<f64>,
    pub daily_fixed_cost: Option<f64>,
    pub region: Option<String>,
}

impl FieldCreateInput {
    pub fn new(name: impl Into<String>, farm_id: i64) -> Self {
        Self {
            name: name.into(),
            farm_id,
            area: None,
            daily_fixed_cost: None,
            region: None,
        }
    }
}
